tests/libmigration/migration.go

```diff
@@ -414,13 +414,13 @@ func ConfirmVMIPostMigrationAborted(vmi *v1.VirtualMachineInstance, migrationUID
 }
 
 func RunMigrationAndExpectFailure(migration *v1.VirtualMachineInstanceMigration, timeout int) string {
-	var err error
 	virtClient := kubevirt.Client()
 	By("Starting a Migration")
 	Eventually(func() error {
-		migration, err = virtClient.VirtualMachineInstanceMigration(migration.Namespace).Create(context.Background(), migration, metav1.CreateOptions{})
+		_, err := virtClient.VirtualMachineInstanceMigration(migration.Namespace).Create(context.Background(), migration, metav1.CreateOptions{})
 		return err
 	}, timeout, 1*time.Second).ShouldNot(HaveOccurred())
```

Comment:
An alternative to this would be to remove the `Eventually()` clause entirely.

I would assume that if the creation fails, it'd probably continue to fail, and the test should fail fast. Is there another example in our test suite for creating an object within an `Eventually`?

WDYT?

Reply:
i guess it's possible for the `Create` to fail due to temporary connectivity issues, i was debugging flaky test in tests/migration/migration.go:1642
in the test flow it managed to create a 7 migrations out of 10, in the 8th iteration the `Create` failed (i don't know what was the original reason for the failure because the error is ignored at first), but the retry logic at this point could succeed if it will use the original migration object, (when i debugged the test the migration object pointed to an empty result struct that resulted en error like this:

Timed out after 180.000s.
Unexpected error:
    <*errors.errorString | 0xc006114b50>: 
    an empty namespace may not be set during creation
    {
        s: "an empty namespace may not be set during creation",
    }

Reply:
> i guess it's possible for the Create to fail due to temporary connectivity issues.

In general, this should not happen. If it does happen, I think it's better for the test to fail in order to investigate what went wrong. My concern here is that this will hide bugs (both production & test related) that would be ignored.

In addition, I don't believe we ever create objects in an Eventually in our tests.

> was debugging flaky test in[ tests/migration/migration.go:1642](https://github.com/kubevirt/kubevirt/blob/8925162860bfd94c4b2e38dd3f035b504cffaee7/tests/migration/migration.go#L1642) in the test flow it managed to create a 7 migrations out of 10, in the 8th iteration the Create failed (i don't know what was the original reason for the failure because the error is ignored at first)

Interesting. Thanks for the detailed explanation.
I wonder if we can reproduce this to better understand the issue. From a first glance, I think it makes sense that the error is due to the fact that the previous migration hasn't been finalized yet, or IOW, there's a race between the test and the controller, since creating two migrations for the same VM is disallowed:
https://github.com/kubevirt/kubevirt/blob/8925162860bfd94c4b2e38dd3f035b504cffaee7/pkg/virt-api/webhooks/validating-webhook/admitters/migration-create-admitter.go#L118-L120

To avoid that we can ensure the migration is finalized before creating another one.
As a side node, I think that creating 10 VMs is a lot and we can probably reduce this number (perhaps to 5?).

Reply:
I agree with @iholder101!
Create should succeed and if not we should see it clearly.
>As a side node, I think that creating 10 VMs is a lot and we can probably reduce this number (perhaps to 5?).

The specific test runs 10 migrations because it want to test that we garbage collect the older migrations. Currently we store the latest 5, so 5 is not enough.
Consider that there the `FuncTestForceLauncherMigrationFailureAnnotation` is used, so the test is fast and there is no lack of resources.
I think that the creation failure is strictly related to https://github.com/kubevirt/kubevirt/pull/13257

Reply:
Done!